### DISCUSSION TITLE
Long range electrostatic, charge fitting

### DISCUSSION THREAD

**Discussion by martino95** [deepmodeling/deepmd-kit#255] [Q&A] [answered]
Hello

I'm working on a model of nucleotides in water using deepmd. Since the system has many charged atoms I think long range electrostatic effects could play a crucial role. I read some articles presenting NN models with long range electrostatic included using a second NN to predict atomic charges. Playing around with your code I found some classes (DipoleChargeModifier, EwaldRecp, DeepDipole) that seem to be here to do this job. When I saw this lines of code in DeepPot.py:
``` Python
            e, f, v = self.eval_inner(coords, cells, atom_types, fparam = fparam, aparam = aparam, atomic = atomic)
            if self.modifier_type is not None:
                me, mf, mv = self.dm.eval(coords, cells, atom_types)
                e += me.reshape(e.shape)
                f += mf.reshape(f.shape)
                v += mv.reshape(v.shape)
            return e, f, v
```
I thought, wow, everything I need is already implemented.
So the question is: am I right? Or those are just pieces of code not intended to be used and left for future development?
If not how can I train a network to predict charges? What's the input data format? What keywords are needed in the .json model file?

Thank you very much

Kindly

Riccardo

**Comment by amcadmus** [comment, marked as answer]
Dear Riccardo,

You are right, we are actively working on including the long-range electrostatics into the deepmd approach. 

The code is not stable at the moment. We will inform you as soon as the it is ready for productive run.

Best regards,
Han

**Reply by KamronF9** [reply]
I am following up with the question above to see where things stand with methods of including charge of atoms in DeePMD during training. In our case, we vary an applied electrostatic potential to a metal-liquid interface in grand canonical (electronic) AIMD and try to learn the behavior as it changes as a function of applied potential. Since charge varies throughout in these systems, charge is needed in the learning process. It appears that the DPLR model now included in DeePMD would learn the charge distribution in the Wannier centroids as described in the Zhang 2022 paper (including applied potential cases). I am not sure it would apply to our system though based on the water examples of the DLPR in your repository. 
Is it required that the atoms have explicit bonds for the deep Wannier model to train?
What would happen if this were used for a non-bonded system (e.g., metal, ionic liquids)? 
Is there another approach you might recommend for the problem we are working on?

I found this too but am not sure how it would be applied. -> "tot_ener_zero: Force the total energy to zero. Useful for the charge fitting." - https://docs.deepmodeling.com/projects/deepmd/en/v2.1.5/_modules/deepmd/fit/ener.html?highlight=charge#

More generally - Is there a way to provide an additional arbitrary conditional input (e.g., applied potential) to the DeePMD training that can be supplied to the training?  That is the training for one potential does work fine.  But including multiple potentials causes conflict in the training and leads to excessive training errors.

I see another Q/A #2136 discussing this same point and see the larger potentials are not yet supported.  What is considered a large potential?

Thanks,

Kamron

**Reply by wanghan-iapcm** [reply]
Dear Kamron,

The DPLR method requires that each Wannier centroid associated to one of the real atoms in the system, thus the method does not applied to the case of charge transfers. Currently we are working on a new approach that handles the cases of charge transfer. 

For ionic liquid, if no charge transfers (no chemical reactions), the DPLR still applies.  For metals, the key question is how to represent the electronic structure of the system in a coarse-grained way. You are welcome to suggest new ideas.

DeePMD-kit can be used to fit atomic charges and predict a configuration dependent partial charge. The `tot_ener_zero` is to neutralize the system in this approach. However, this is not the suggest way of modeling long-range interactions, because the partial charge is not a physically well defined observable, and the error introduced does not converge with respect to the cut-off radius. 

In #2136, the large field refers to the field that is outside the linear response regime. The magnitude is system dependent.  

Best,
Han

**Reply by hanao2** [reply]
Hello Han,

I have a question about the DPLR model. Suppose I have an electrified interface but the system remains in the non-Faradaic region (only molecular reconfiguration takes place and there is no chemical reaction). The DPLR model should be applicable to this system, right?

Thanks!

**Reply by wanghan-iapcm** [reply]
Yes DPLR should apply.
